### PR TITLE
Add IoCContainer & ToolsRegistry

### PR DESCRIPTION
**Pull request by gohabereg**

### CODE REVIEW

**Comment by gohabereg** [comment]
resolves #35

File: src/tools/types/DataType.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Possible types of BlockData values
+ */
+export enum DataType {
```
Comment (neSpecc) [inline src/tools/types/DataType.ts:4]: I'd suggest `BlockDataType`

File: src/tools/types/BlockToolDataScheme.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { DataType } from './DataType';
+import { DataKey } from '../../entities';
+
+/**
+ * Block Tool data scheme to describe the structure of the data
+ */
+export type BlockToolDataScheme = Record<DataKey, DataType>;
```
Comment (neSpecc) [inline src/tools/types/BlockToolDataScheme.ts:7]: What if Tool contains TextNodes not on the first level, like NestedList or Table tool:

```
data: {
  items: [
    {
      content: ''
      items: [...]
    }
  ]
}
```

**Review by ilyamore88 — APPROVED** [review]

File: src/IoC/Container.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import type { EditorDocument } from '../entities';
+
+type Container = Map<symbol, unknown>;
+
+/**
+ * Singletone IoC container to store EditorDocuments dependencies, for example ToolsRegistry and EventBus
```
Comment (neSpecc) [inline src/IoC/Container.ts:6]: pls describe why do we need this in comments

File: src/IoC/Container.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import type { EditorDocument } from '../entities';
+
+type Container = Map<symbol, unknown>;
```
Comment (neSpecc) [inline src/IoC/Container.ts:3]: add doc here pls

File: src/IoC/Container.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * const registry = container.get(TOOLS_REGISTRY);
+ * ```
+ */
+export class IoCContainer {
+  /**
+   * Map of containers for each document
+   *
+   * @private
+   */
+  static #containers: Map<EditorDocument, Container> = new Map();
```
Comment (neSpecc) [inline src/IoC/Container.ts:27]: maybe `WeakMap` to allow EditorDocument be GCed?

**Review by neSpecc — APPROVED** [review]

**Pull request merged**